### DISCUSSION TITLE
How to use a DeflateConfig?

### DISCUSSION THREAD

**Discussion by gurinderu** [libp2p/rust-libp2p#4196] [Q&A]
There are no examples of how to use a deflate config. Could you provide a way or probably add another way to add compression on the connection level or stream level?

**Comment by mxinden** [comment]
You can use it at the `ConnectionHandler` level when upgrading a stream. E.g. on `ConnectionHandler::listen_protocol` you can return an `InboundUpgrade` that first negotiates your application protocol and then uses `libp2p_core::upgrade::apply` to also apply the deflate upgrade.

All that said, I am not aware of anyone using it. In other words I am not aware of any use-case where it has proven useful thus far. See also https://research.protocol.ai/blog/2020/honey-i-shrunk-our-libp2p-streams/.
